npm_and_yarn/lib/dependabot/npm_and_yarn/version.rb

```diff
@@ -34,6 +34,27 @@ def initialize(version)
         super
       end
 
+      def major
+        @major ||= segments[0] || 0
+      end
+
+      def minor
+        @minor ||= segments[1] || 0
+      end
+
+      def patch
+        @patch ||= segments[2] || 0
+      end
+
+      def backwards_compatible_with?(other)
+        case major
+        when 0
+          self == other
```

Comment:
According to the [SemVer spec](https://semver.org)

> 4. Major version zero (0.y.z) is for initial development. Anything MAY change at any time. The public API SHOULD NOT be considered stable.

> 6. Patch version Z (x.y.Z | x > 0) MUST be incremented if only backwards compatible bug fixes are introduced. A bug fix is defined as an internal change that fixes incorrect behavior.

I can't tell if the carve-out for `x > 0` means that we should match with strict equality here. Should we instead do this?

```suggestion
          major == other.major && minor == other.minor
```

Reply:
Then again, npm advises releases start numbering at `1.0.0`, and there only appears to be a [single `@types` package with a major zero version](https://www.npmjs.com/search?q=%40types%200.) So maybe this is fine as is?

Reply:
Dependabot's [version update docs](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/about-dependabot-version-updates) don't promise compatibility (emphasis mine):

> Dependabot takes the effort out of maintaining your dependencies. You can use it to ensure that your repository automatically keeps up with **the latest releases** of the packages and applications it depends on. That's sort of a different concern though (direct dependencies).

We *should* ship compatible packages together. Given items `4` and `6` from the SemVer spec, I think the current code is the only one in which we can guarantee backward compatibility.

---

💡 I wonder if this method would benefit from being a little more specific in its name or argument. In our specific use case, we're considering two packages backward compatible if their versioning schemes could safely resolve to each others. I don't think that's necessarily true of other packages (like peer dependencies). I'll try to think of something and update this comment if I do 😄